### ISSUE TITLE
How to change the LSTMCell weight format from tensorflow to tf.keras

### ISSUE THREAD

**Issue by 5M7k0R5tTe** [tensorflow/tensorflow#29622]
I I have some old code from tensorflow that I want to make work for tensorflow2/tf.keras. I would like to keep the same LSTM weights, but cannot figure out how to convert the format.

I have the old weights saved in a checkpoint file, and also have them saved in csv files.

My old code looks something like this:
```
input_placeholder = tf.placeholder(tf.float32, [None, None, input_units])
lstm_layers = [tf.nn.rnn_cell.LSTMCell(layer_size), tf.nn.rnn_cell.LSTMCell(layer_size)]
stacked = tf.contrib.rnn.MultiRNNCell(lstm_layers)
features, state = tf.nn.dynamic_rnn(stacked, input_placeholder, dtype=tf.float32)
```

And my new code looks something like this:
```
input_placeholder = tf.placeholder(tf.float32, [None, None, input_units])
lstm_layers = [tf.keras.layers.LSTMCell(layer_size),tf.keras.layers.LSTMCell(layer_size)]
stacked = tf.keras.layers.StackedRNNCells(lstm_layers)
features = stacked(input_placeholder)
... #later in the code
features.set_weights(previous_weights)
```

The old bias seems to match the new bias. The old kernel seems to be the concatenation of the kernel and recurrent kernel. I am able to load the previous_weights into the model (have explicitly checked the weights loaded correctly), however tests I have fail to produce the same result. Digging into the source code, the kernels seem to have a different format under the hood.

Is it possible to calculate the kernel and recurrent_kernel (tf.keras) using these old saved kernel weights?

Links if they're helpful:

https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/ops/rnn_cell_impl.py

https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/keras/layers/recurrent.py

**Comment by jvishnuvardhan** [comment]
@5M7k0R5tTe Please fill the issue [template](https://github.com/tensorflow/tensorflow/issues/new?template=10-build-installation-issue.md). Could you update them if they are relevant in your case, or leave them as N/A? Along with the template, please provide as many details as possible to find the root cause of the issue. Thanks!

**Comment by qlzh727** [comment]
@5M7k0R5tTe, the tf lstm cell has two weights: kernel, bias. and keras cell has 3 weights: kernel, recurrent_kernel, bias. The tf kernel should be the concat of keras kernel and recurrent kernel. 

Format wise, tf lstm cell was using IFCO (input, forget, carry, output) gate order, and Keras cell was using ICFO (input, carry, forget, output). So to convert the weights, assume each gates has size U, it will be:

i, f, c, o = np.split(tf_weights, 4, axis=1)
keras_weights = np.concatenate([i, c, f, o], axis=1)

**Comment by q-ycong-p** [comment]
@qlzh727 , thanks for the clarification. However, it almost looks like TF LSTMCell uses icfo, and keras LSTM uses ifco, as opposite to what you described?

The source is: current tf2onnx handles TF LSTMCell (see code [here](https://github.com/onnx/tensorflow-onnx/blob/master/tests/test_lstm.py#L20)) and the lstm rewriter assumes that the [weights](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_rewriter.py#L222) and [bias](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_rewriter.py#L224) are in icfo, prior to conversion to ONNX format, which is [iofc](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_rewriter.py#L235). 

Please correct me if wrong. I am confused after reading tf2onnx code and this thread. Thanks.

**Comment by qlzh727** [comment]
oh, yes. Thanks for spotting the error in my previous message. The keras weight format is IFCO, and TF format was ICFO. 

You can find the keras logic for weights in https://github.com/keras-team/keras/blob/353cb10af7c0c7d435b31129c5970fcabe3fc961/keras/layers/recurrent.py#L2468

> @qlzh727 , thanks for the clarification. However, it almost looks like TF LSTMCell uses **icfo**, and keras LSTM uses **ifco**, as opposite to what you described?
> 
> The source is: current tf2onnx handles TF LSTMCell (see code [here](https://github.com/onnx/tensorflow-onnx/blob/master/tests/test_lstm.py#L20)) but [not keras LSTM as of now](https://github.com/onnx/tensorflow-onnx/issues/1546). Under this case, the lstm rewriter treats that the [weights](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_rewriter.py#L222) and [bias](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_rewriter.py#L224) as in **icfo** format. Only then, it's converted to ONNX format which is [**iofc**](https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/rewriter/lstm_rewriter.py#L235).
> 
> Therefore, my understanding on default gate order is: TF LSTM: ICFO keras LSTM: IFCO ONNX LSTM: IOFC
> 
> Please correct me if wrong. I am confused after reading tf2onnx code and this thread. Thanks.
